### PR TITLE
Attempt reconnect from startup

### PR DESCRIPTION
**Pull request by jovdipp**

### CODE REVIEW

File: src/ecpool_worker.erl [hunk 1]
```diff
@@ -211,8 +214,11 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal Functions
 %%--------------------------------------------------------------------
 
-connect(#state{mod = Mod, opts = Opts, id = Id}) ->
-    Mod:connect([{ecpool_worker_id, Id} | connopts(Opts, [])]).
+connect(#state{mod = Mod, opts = Opts, id = Id, initiated = Initiated,
+               connection_attempts = Attempts}) ->
+    ConOpts = [ {ecpool_worker_id, Id}, {initiated, Initiated}, {connection_attempts, Attempts}
```
Comment (zmstone) [inline src/ecpool_worker.erl:219]: why pass down internal flags to callback ?
unknown options may cause unexpected behaviour in the callback

for testing ? maybe call sys:get_state to inspect the internal state.
or use https://github.com/kafka4beam/snabbkaffe

File: src/ecpool_worker.erl [hunk 2]
```diff
@@ -226,11 +232,29 @@ connopts([{Key, _}|Opts], Acc)
 connopts([Opt|Opts], Acc) ->
     connopts(Opts, [Opt|Acc]).
 
-reconnect(Secs, State = #state{client = Client, on_disconnect = Disconnect, supervisees = SubPids}) ->
+
+init_reconnect(#state{opts = Opts} = State, Error) ->
```
Comment (zmstone) [inline src/ecpool_worker.erl:236]: Maybe we don't need this. The only case to return `stop` from here is when attempt limit is 1.

I'd suggest in `init/1`, it simply sends a loopback message to start the init-connect retry loop.
i.e. `init/1` always succeeds immediately so it does not slow down its supervisor.

So, the `handle_info(reconnect` clause becomes the only entry-point where connect attempts are made.

**Review by zmstone — COMMENTED** [review]
The new option `start_up_attempts` should be filtered by `connect_opts/2`.
Otherwise it get passed down to Mod:connect.

File: src/ecpool_worker.erl [hunk 3]
```diff
@@ -48,6 +48,9 @@
           id :: pos_integer(),
           client :: pid() | undefined,
           mod :: module(),
+          initiated = false,
```
[the remark this answers is not in the document]
Comment (jovdipp) [reply src/ecpool_worker.erl:51]: Yeah

File: src/ecpool_worker.erl [hunk 4]
```diff
@@ -226,11 +232,29 @@ connopts([{Key, _}|Opts], Acc)
...
+    Secs = proplists:get_value(auto_reconnect, Opts, false),
+    case reconnect(Secs, State) of
+        {stop, AttemptsError} ->
+            ErrorWithCause = {AttemptsError, Error},
+            {stop, ErrorWithCause};
+        {noreply, StateUpdated} ->
+            {ok, StateUpdated}
+    end.
+
+reconnect(_Secs, #state{initiated = false, connection_attempts = Attempts,
```
[the remark this answers is not in the document]
Comment (jovdipp) [reply src/ecpool_worker.erl:246]: I set the default in the config schema to 1 if not set.

File: src/ecpool_worker.erl (hunk 1 above)
Comment (jovdipp) [reply src/ecpool_worker.erl:219]: The two flags are sent for the users benefit, it's necessary to allow the callbacks to make better connect behaviour decisions.

File: src/ecpool_worker.erl (hunk 2 above)
Comment (jovdipp) [reply src/ecpool_worker.erl:236]: It already does, that reconnect simply sends the "reconnect" message.

File: src/ecpool_worker.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (jovdipp) [reply src/ecpool_worker.erl:246]: I set a default in the auth_mysql schema as 1, which is obviously not the only app using ecpool :)
Will correct it

File: src/ecpool_worker.erl (hunk 1 above)
Comment (zmstone) [reply src/ecpool_worker.erl:219]: in that case, maybe namespace them, e.g. `ecpool_is_retry_attempt` and  `ecpool_connection_attempt` ?

Comment (zmstone) [reply src/ecpool_worker.erl:219]: or maybe this: `{ecpool_connect_type, initial | initial_retry | normal_retry}`

File: src/ecpool_worker.erl (hunk 2 above)
Comment (zmstone) [reply src/ecpool_worker.erl:236]: I mean if we treat connect and reconnect equal, there is no need to handle a 'stop after first attempt' special case.

Comment (jovdipp) [reply src/ecpool_worker.erl:236]: I agree, although did that way to try and not affect any current expectant behavior. I assume ec_pool is widely used.
Ordinarily I would say if the tests pass then all should be good until something break and a new test is required

Comment (zmstone) [reply src/ecpool_worker.erl:236]: fair.

**Comment by zmstone** [comment]
maybe in another PR: we need github action for the tests.

**Pull request closed without merging**